model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/locking/HotRodGlobalLockProvider.java

```diff
@@ -0,0 +1,87 @@
+
+import org.infinispan.client.hotrod.MetadataValue;
+import org.infinispan.client.hotrod.RemoteCache;
+import org.jboss.logging.Logger;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.locking.GlobalLock;
+import org.keycloak.models.locking.GlobalLockProvider;
+import org.keycloak.models.locking.LockAcquiringTimeoutException;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.LinkedList;
+import java.util.List;
+
+
+public class HotRodGlobalLockProvider implements GlobalLockProvider {
+
+    private static final Logger LOG = Logger.getLogger(HotRodGlobalLockProvider.class);
+    private final RemoteCache<String, String> locksCache;
+    private final long defaultTimeoutMilliseconds;
+    private final List<GlobalLock> providedGlobalLocks = new LinkedList<>();
+
+    public HotRodGlobalLockProvider(KeycloakSession session, RemoteCache<String, String> locksCache, long defaultTimeoutMilliseconds) {
+        this.locksCache = locksCache;
+        this.defaultTimeoutMilliseconds = defaultTimeoutMilliseconds;
+
+        // Enlist this provider for closing so all provided locks are closed if not closed by AutoClosable after commit
+        session.enlistForClose(this);
+    }
+
+    @Override
+    public GlobalLock acquire(String lockName, Duration timeToWaitForLock) throws LockAcquiringTimeoutException {
+        if (timeToWaitForLock == null) {
+            // Set default timeout if null provided
+            timeToWaitForLock = Duration.ofMillis(defaultTimeoutMilliseconds);
+        }
+
+        LOG.debugf("Acquiring lock [%s].", lockName);
+        if (!HotRodLocksUtils.repeatPutIfAbsent(locksCache, lockName, timeToWaitForLock.toMillis(), 50)) {
+            MetadataValue<String> withMetadata = locksCache.getWithMetadata(lockName);
+            throw new LockAcquiringTimeoutException(lockName, withMetadata.getValue(), Instant.ofEpochMilli(withMetadata.getCreated()));
+        }
+        LOG.debugf("Lock acquired [%s].", lockName);
+        GlobalLock lock = new HotRodGlobalLock(() -> {
+            LOG.debugf("Releasing lock [%s].", lockName);
+            boolean result = HotRodLocksUtils.removeWithInstanceIdentifier(locksCache, lockName);
+            LOG.debugf("Lock [%s] release resulted with %s", lockName, result);
+
+            if (!result) {
+                LOG.errorf("Unlocking the lock [%s] was unsuccessful. Current lock value: [%s]", lockName, locksCache.get(lockName));
+                throw new IllegalStateException("Unlocking the lock was unsuccessful. See log for more details.");
+            }
+        });
+
+        // Keep reference to acquired lock, so it can be closed by on session close
+        providedGlobalLocks.add(lock);
+        return lock;
+    }
+
+    @Override
+    public void forceReleaseAllLocks() {
+        locksCache.clear();
+    }
+
+    @Override
+    public void close() {
+        providedGlobalLocks.forEach(GlobalLock::close);
+    }
+}
```

Comment:
Good point @ahus1! I think it makes sense to do the cleanup as is now, at the session cleanup because we sometimes `enlistAfterCompletion` the map storage transaction and it makes sense to clean locks only after it commits. On the other hand, having it part of the main transaction reads better. WDYT?